### ISSUE TITLE
ValueError: Unknown column geometry

### ISSUE THREAD

**Issue by palaiya** [geopandas/geopandas#1144]
I am getting the following exception on `mac os 10.14.5` despite my condarc has strict mode.  

**Error:**
<details>

```
Error : Traceback (most recent call last):
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2897, in get_loc
    return self._engine.get_loc(key)
  File "pandas/_libs/index.pyx", line 107, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 131, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'geometry'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 155, in set_geometry
    level = frame[col].values
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 546, in __getitem__
    result = super(GeoDataFrame, self).__getitem__(key)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/pandas/core/frame.py", line 2980, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2899, in get_loc
    return self._engine.get_loc(self._maybe_cast_indexer(key))
  File "pandas/_libs/index.pyx", line 107, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 131, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'geometry'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "ImageCoregister.py", line 18, in <module>
    a = CRL.correct_shifts()
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 603, in correct_shifts
    coreg_info = self.coreg_info
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 575, in coreg_info
    'GCPList': self.tiepoint_grid.GCPList,
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 333, in tiepoint_grid
    self._tiepoint_grid.get_CoRegPoints_table()
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/Tie_Point_Grid.py", line 402, in get_CoRegPoints_table
    GDF = GDF.astype(np.object).merge(records.astype(np.object), on='POINT_ID', how="inner")
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 711, in astype
    df = geopandas.GeoDataFrame(df, geometry=self._geometry_column_name)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 83, in __init__
    self.set_geometry(geometry, inplace=True)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 157, in set_geometry
    raise ValueError("Unknown column %s" % col)
ValueError: Unknown column geometry

```
</details>

**conda list**

<details>

```

affine                    2.3.0                      py_0    conda-forge
arosics                   0.9.2                    pypi_0    pypi
asciitree                 0.3.3                      py_2    conda-forge
asn1crypto                1.0.0                    py36_0    conda-forge
attrs                     19.2.0                     py_0    conda-forge
basemap                   1.2.1            py36hd103db2_1    conda-forge
bokeh                     1.3.4                    py36_0    conda-forge
boost-cpp                 1.70.0               hd59e818_1    conda-forge
branca                    0.3.1                    pypi_0    pypi
bzip2                     1.0.8                h01d97ff_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
cairo                     1.16.0            h0ab9d94_1001    conda-forge
certifi                   2019.6.16                py36_1    conda-forge
cffi                      1.12.3           py36hccf1714_0    conda-forge
cfitsio                   3.470                h389770f_2    conda-forge
chardet                   3.0.4                 py36_1003    conda-forge
click                     7.0                        py_0    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
cligj                     0.5.0                      py_0    conda-forge
cloudpickle               1.2.2                      py_0    conda-forge
cmocean                   2.0                        py_1    conda-forge
colorspacious             1.1.2              pyh24bf2e0_0    conda-forge
configobj                 5.0.6                      py_0    conda-forge
cryptography              2.7              py36h212c5bf_0    conda-forge
curl                      7.65.3               h22ea746_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
cytoolz                   0.10.0           py36h01d97ff_0    conda-forge
dask                      2.5.0                      py_0    conda-forge
dask-core                 2.5.0                      py_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
dill                      0.3.1.1                  pypi_0    pypi
distributed               2.5.1                      py_0    conda-forge
earthpy                   0.7.4                      py_0    conda-forge
expat                     2.2.5             h6de7cb9_1003    conda-forge
fasteners                 0.14.1                     py_3    conda-forge
fiona                     1.8.6            py36h7eb7563_3    conda-forge
folium                    0.10.0                   pypi_0    pypi
fontconfig                2.13.1            h1027ab8_1000    conda-forge
freetype                  2.10.0               h24853df_1    conda-forge
freexl                    1.0.5             h1de35cc_1002    conda-forge
fsspec                    0.5.1                      py_0    conda-forge
gdal                      2.4.1           py36h39889d8_10    conda-forge
geoarray                  0.8.22                   pypi_0    pypi
geojson                   2.5.0                    pypi_0    pypi
geopandas                 0.6.0                      py_0    conda-forge
geos                      3.7.2                h6de7cb9_2    conda-forge
geotiff                   1.4.3             h113155d_1001    conda-forge
gettext                   0.19.8.1          h46ab8bc_1002    conda-forge
giflib                    5.1.7                h01d97ff_1    conda-forge
glib                      2.58.3            h9d45998_1002    conda-forge
hdf4                      4.2.13            hf3c6af0_1002    conda-forge
hdf5                      1.10.4          nompi_h0cbb7df_1106    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       58.2              h0a44026_1000    conda-forge
idna                      2.8                   py36_1000    conda-forge
imageio                   2.5.0                    py36_0    conda-forge
jinja2                    2.10.1                     py_0    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
json-c                    0.13.1            h1de35cc_1001    conda-forge
kealib                    1.4.10            hecf890f_1003    conda-forge
kiwisolver                1.1.0            py36h770b8ee_0    conda-forge
krb5                      1.16.3            hcfa6398_1001    conda-forge
libblas                   3.8.0               13_openblas    conda-forge
libcblas                  3.8.0               13_openblas    conda-forge
libcurl                   7.65.3               h16faf7d_0    conda-forge
libcxx                    9.0.0                         0    conda-forge
libcxxabi                 9.0.0                         0    conda-forge
libdap4                   3.20.4               habf5908_0    conda-forge
libedit                   3.1.20170329      hcfe32e1_1001    conda-forge
libffi                    3.2.1             h6de7cb9_1006    conda-forge
libgdal                   2.4.1               hc4192b8_10    conda-forge
libgfortran               3.0.1                         0    conda-forge
libiconv                  1.15              h01d97ff_1005    conda-forge
libkml                    1.3.0             hed7d534_1010    conda-forge
liblapack                 3.8.0               13_openblas    conda-forge
libnetcdf                 4.6.2             h6b88ef6_1001    conda-forge
libopenblas               0.3.7                hd44dcd8_1    conda-forge
libpng                    1.6.37               h2573ce8_0    conda-forge
libpq                     11.5                 h756f0eb_1    conda-forge
libspatialindex           1.9.0                h6de7cb9_1    conda-forge
libspatialite             4.3.0a            hb3e03d6_1030    conda-forge
libssh2                   1.8.2                hcdc9a53_2    conda-forge
libtiff                   4.0.10            hd08fb8f_1003    conda-forge
libxml2                   2.9.9                hd80cff7_2    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
markupsafe                1.1.1            py36h1de35cc_0    conda-forge
matplotlib                3.1.1                    py36_1    conda-forge
matplotlib-base           3.1.1            py36h3a684a6_1    conda-forge
monotonic                 1.5                        py_0    conda-forge
mpld3                     0.3                      pypi_0    pypi
msgpack-python            0.6.2            py36h770b8ee_0    conda-forge
munch                     2.3.2                      py_0    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
networkx                  2.3                        py_0    conda-forge
numcodecs                 0.6.3            py36h0a44026_0    conda-forge
numpy                     1.17.2           py36h6b0580a_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openjpeg                  2.3.1                hdc36067_1    conda-forge
openssl                   1.1.1c               h01d97ff_0    conda-forge
packaging                 19.2                       py_0    conda-forge
pandas                    0.25.1           py36h86efe34_0    conda-forge
partd                     1.0.0                      py_0    conda-forge
pcre                      8.41              h0a44026_1003    conda-forge
pillow                    6.2.0            py36hb6f49c9_0    conda-forge
pip                       19.2.3                   py36_0    conda-forge
pixman                    0.38.0            h01d97ff_1003    conda-forge
plotly                    4.1.1                    pypi_0    pypi
poppler                   0.67.0               hd5eb092_7    conda-forge
poppler-data              0.4.9                         1    conda-forge
postgresql                11.5                 h25afefd_1    conda-forge
proj4                     5.2.0             h6de7cb9_1006    conda-forge
psutil                    5.6.3            py36h01d97ff_0    conda-forge
py-tools-ds               0.14.21                  pypi_0    pypi
pycparser                 2.19                     py36_1    conda-forge
pyfftw                    0.11.1          py36h917ab60_1001    conda-forge
pykdtree                  1.3.1           py36h917ab60_1002    conda-forge
pykrige                   1.4.0           py36h1d22016_1001    conda-forge
pyopenssl                 19.0.0                   py36_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyproj                    1.9.6           py36h01d97ff_1002    conda-forge
pyresample                1.13.1           py36h86efe34_0    conda-forge
pyshp                     2.1.0                      py_0    conda-forge
pysocks                   1.7.1                    py36_0    conda-forge
python                    3.6.7             h94afb7f_1005    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
pytz                      2019.2                     py_0    conda-forge
pywavelets                1.0.3            py36h1e5eb4f_1    conda-forge
pyyaml                    5.1.2            py36h01d97ff_0    conda-forge
rasterio                  1.0.25           py36h238e834_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
requests                  2.22.0                   py36_1    conda-forge
retrying                  1.3.3                    pypi_0    pypi
rtree                     0.8.3           py36h666c49c_1002    conda-forge
scikit-image              0.15.0           py36h86efe34_2    conda-forge
scipy                     1.3.1            py36hab3da7d_2    conda-forge
setuptools                41.2.0                   py36_0    conda-forge
shapely                   1.6.4           py36h0567c5e_1006    conda-forge
six                       1.12.0                py36_1000    conda-forge
snuggs                    1.4.7                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
spectral                  0.19                     pypi_0    pypi
sqlite                    3.29.0               hb7d70f7_1    conda-forge
tblib                     1.4.0                      py_0    conda-forge
tk                        8.6.9             h2573ce8_1003    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.3            py36h01d97ff_0    conda-forge
tzcode                    2019a             h01d97ff_1002    conda-forge
urllib3                   1.25.6                   py36_0    conda-forge
utm                       0.5.0                    pypi_0    pypi
viscm                     0.7                pyh24bf2e0_0    conda-forge
wheel                     0.33.6                   py36_0    conda-forge
xerces-c                  3.2.2             h4db8090_1003    conda-forge
xz                        5.2.4             h1de35cc_1001    conda-forge
yaml                      0.1.7             h1de35cc_1001    conda-forge
zarr                      2.3.2                    py36_0    conda-forge
zict                      1.0.0                      py_0    conda-forge
zlib                      1.2.11            h01d97ff_1006    conda-forge
zstd                      1.4.0                ha9f0a20_0    conda-forge

```
</details>

**condarc**  

```
auto_activate_base: false
ssl_verify: true
channels:
  - conda-forge
  - defaults
channel_priority: strict
```

**conda info :**

   ```
active environment : dev
    active env location : /Users/Naresh/anaconda3/envs/dev
            shell level : 1
       user config file : /Users/Naresh/.condarc
 populated config files : /Users/Naresh/.condarc
          conda version : 4.7.12
    conda-build version : 3.18.8
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /Users/Naresh/anaconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/Naresh/anaconda3/pkgs
                          /Users/Naresh/.conda/pkgs
       envs directories : /Users/Naresh/anaconda3/envs
                          /Users/Naresh/.conda/envs
               platform : osx-64
             user-agent : conda/4.7.12 requests/2.22.0 CPython/3.7.3 Darwin/18.6.0 OSX/10.14.5
                UID:GID : 501:20
             netrc file : /Users/Naresh/.netrc
           offline mode : False
```

**Comment by brendan-ward** [comment]
@palaiya I'm not sure if it is helpful based on your situation above, but I recently encountered s similar issue with the 0.6.0 version.  I had a GeoDataFrame, from which I did an operation that extracted a few of the non-geometry columns.  Under previous versions, this would have been cast as a DataFrame, not GeoDataFrame, due to the lack of a geometry column.  However, this time it still considered itself a GeoDataFrame, and on subsequent operations (e.g., `.copy()` or returning it from a function), it failed due to the missing geometry column.

To see if this is the case, look at the `type` of your data frame just before the error is raised; if it is a `GeoDataFrame` but should be a `DataFrame` because it lacks a geometry column, then just make sure to cast it back to a `DataFrame` before doing subsequent operations.

**Comment by palaiya** [comment]
@brendan-ward I am just trying to run [arosics example](http://danschef.gitext.gfz-potsdam.de/arosics/doc/usage/local_coreg.html#detect-and-correct-local-shifts-with-input-data-on-disk) and it gives the error as i pasted above.

**Comment by jorisvandenbossche** [comment]
@brendan-ward do you have a reproducible example of what you describe? 
Because we should see if we need to fix something for that in 0.6.1, as that sounds as a regression

**Comment by palaiya** [comment]
@jorisvandenbossche  As i said i was trying to co-register two images using this [library and example](http://danschef.gitext.gfz-potsdam.de/arosics/doc/usage/local_coreg.html#detect-and-correct-local-shifts-with-input-data-on-disk) 

```
from arosics import COREG_LOCAL

im_reference = '/path/to/your/ref_image.bsq'
im_target    = '/path/to/your/tgt_image.bsq'
kwargs = {
    'grid_res'     : 200,
    'window_size'  : (64,64),
    'path_out'     : 'auto',
    'q'            : False,
    'fmt_out'      : 'GTIFF',
 }

CRL = COREG_LOCAL(im_reference,im_target,**kwargs)
a = CRL.correct_shifts()
```

Now to run the above example. you will need to install all the dependencies as mentioned [here](http://danschef.gitext.gfz-potsdam.de/arosics/doc/installation.html) in the fresh conda environment. Once you install `arosics` and run the example code you will be able to get the issue.

**Comment by jorisvandenbossche** [comment]
OK, took a quick look, and found the following issues.

If you create a GeoDataFrame without actual geometries (which we still allow), and then do `astype` on the full dataframe, this fails (while this worked before):

```
In [17]: gdf = geopandas.GeoDataFrame({'a': [0, 1]}) 

In [18]: gdf.astype(object) 
...
ValueError: Unknown column geometry
```

So this is something we can fix in geopandas. 
However, it is also `arosics` that is creating a GeoDataFrame that should be a plain DataFrame (https://gitext.gfz-potsdam.de/danschef/arosics/blob/master/arosics/Tie_Point_Grid.py#L396-399), so that could also be updated there.

Additionally, I also noticed that `astype(object)` on a proper GeoDataFrame preserves the GeoDataFrame class:

```
In [19]: gdf = geopandas.GeoDataFrame({'a': [0, 1], 'geometry': geopandas.points_from_xy([1, 2],[1, 2])})  

In [20]: gdf.astype(object)
Out[20]: 
   a                 geometry
0  0  POINT (1.00000 1.00000)
1  1  POINT (2.00000 2.00000)

In [21]: type(gdf.astype(object)) 
Out[21]: geopandas.geodataframe.GeoDataFrame
```

while before 0.6.0, this returned a normal DataFrame. 
While the above is still a valid GeoDataFrame, in theory we should probably convert to a normal DataFrame with an `object` dtype column instead of preserving a `geometry` dtype column as is done now.

@palaiya thanks for opening the issue, BTW !

**Comment by jorisvandenbossche** [comment]
@brendan-ward  so I found the cause of the original reported issue, but if you can still find a reproducible example of what you described in https://github.com/geopandas/geopandas/issues/1144#issuecomment-539070783, that would be very welcome. 

I can't directly reproduce it. For example in:

```
In [25]: df = geopandas.read_file(geopandas.datasets.get_path('naturalearth_lowres'))

In [26]: type(df) 
Out[26]: geopandas.geodataframe.GeoDataFrame

In [27]: subset = df[['continent', 'name']]  

In [28]: type(subset)
Out[28]: pandas.core.frame.DataFrame
```

taking a subset of non-geometry columns correctly returns a DataFrame.

**Comment by brendan-ward** [comment]
@jorisvandenbossche Thanks for investigating this.  I think I found the issue.  Using your `df` above:

Using `.loc` to filter and also extract specific columns:
```
>>> type(df.loc[df.name == 'Fiji', ['continent']])
<class 'geopandas.geodataframe.GeoDataFrame'>
```
Even though it has no `geometry` column.

Whereas using `.loc` to filter first and then extracting a subset of columns works fine.
```
>>> type(df.loc[df.name == 'Fiji'][['continent']])
<class 'pandas.core.frame.DataFrame'>
```

In my specific case, I was using the above example of `.loc` to extract out specific columns from a GeoDataFrame (no longer needed the geometry from it), then immediately joining it to another DataFrame, and I wanted a DataFrame as output.

**Comment by jorisvandenbossche** [comment]
@brendan-ward ah, yes, I see. Thanks for the example!

Hmm, that's more difficult to fix on the geopandas side, as we don't control `loc`. Although I suppose we could do something similar as the `GeoSeries` constructor, where we have a version that falls back to plain DataFrame if there are no geometries: https://github.com/geopandas/geopandas/blob/da3aef6d4bbf0b2308aaeed36e0b5904fd7526ec/geopandas/geoseries.py#L31-L47

If we do something similar for `GeoDataFrame._constructor`, I suppose that this might fix that issue.

**Comment by jorisvandenbossche** [comment]
@brendan-ward this issue is now fixed on master, but I opened https://github.com/geopandas/geopandas/issues/1153 with your description of the `.loc` issue

**Comment by danschef** [comment]
> Traceback (most recent call last):
  File "ImageCoregister.py", line 18, in <module>
    a = CRL.correct_shifts()
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 603, in correct_shifts
    coreg_info = self.coreg_info
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 575, in coreg_info
    'GCPList': self.tiepoint_grid.GCPList,
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/CoReg_local.py", line 333, in tiepoint_grid
    self._tiepoint_grid.get_CoRegPoints_table()
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/arosics/Tie_Point_Grid.py", line 402, in get_CoRegPoints_table
    GDF = GDF.astype(np.object).merge(records.astype(np.object), on='POINT_ID', how="inner")
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 711, in astype
    df = geopandas.GeoDataFrame(df, geometry=self._geometry_column_name)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 83, in __init__
    self.set_geometry(geometry, inplace=True)
  File "/Users/Naresh/anaconda3/envs/dev/lib/python3.6/site-packages/geopandas/geodataframe.py", line 157, in set_geometry
    raise ValueError("Unknown column %s" % col)
ValueError: Unknown column geometry

> However, it is also arosics that is creating a GeoDataFrame that should be a plain DataFrame (https://gitext.gfz-potsdam.de/danschef/arosics/blob/master/arosics/Tie_Point_Grid.py#L396-399), so that could also be updated there.

This is fixed in version 0.9.3 of AROSICS.
